jobbergate-core/jobbergate_core/auth/core.py

```diff
@@ -0,0 +1,316 @@
+
+        Returns:
+            Token: The token is returned for reference.
+
+        Raises:
+            AuthenticationError: If the credentials are invalid.
+        """
+        AuthenticationError.require_condition(token_type in self.tokens, f"{token_type} token was not found")
+        token = self.tokens[token_type]
+        AuthenticationError.require_condition(not token.is_expired(), f"{token_type} token has expired")
+
+        return token
+
+    def load_from_cache(self, skip_loaded: bool = True):
+        """
+        Load the tokens that are available at the cache directory.
+
+        Arguments:
+            skip_loaded (bool): If True, skip tokens that are already loaded, otherwise
+                replace them with the ones from the cache.
+
+        Note:
+            This method does not check if any required token is missing nor is expired,
+            for that, see :meth:`JobbergateAuth.check_credentials`.
+        """
+        logger.debug("Loading tokens from cache directory: {}", self.cache_directory.as_posix())
+
+        for t in TokenType:
+            if t in self.tokens and skip_loaded:
+                continue
+            try:
+                new_token = Token.load_from_cache(self.cache_directory, label=t)
+                self.tokens[t] = new_token
+            except TokenError as e:
+                logger.debug(f"   Error while loading {t}.token: {str(e)}")
+
+    def save_to_cache(self):
+        """
+        Save the tokens to the cache directory.
+
+        Note:
+            This method will create the cache directory if it does not exist.
+        """
+        logger.debug(
+            "Saving tokens to cache directory: {}",
+            self.cache_directory.as_posix(),
+        )
+        self.cache_directory.mkdir(parents=True, exist_ok=True)
+        for token in self.tokens.values():
+            token.save_to_cache()
+
+    def logout(self):
+        """
+        Logout from Jobbergate by clearing the loaded tokens and their cache on the disk.
+        """
+        logger.debug("Logging out from Jobbergate")
+        for token in self.tokens.values():
+            token.clear_cache()
+        self.tokens.clear()
+
+    def login(self):
+        """
+        Login to Jobbergate.
+
+        An URL will be printed to the console, the user must open it in a browser and provide
+        their access credentials.
+
+        After the login is completed, the tokens will be saved to the cache directory.
+        """
+        logger.debug("Preparing to login to Jobbergate")
+        login_info = self._get_login_information()
+        response = self._wait_for_login_confirmation(login_info)
+        self._process_tokens_from_response(response)
+        logger.success("Login completed")
+
+    def _wait_for_login_confirmation(self, login_info: _LoginInformation) -> httpx.Response:
+        print(f"Login Here: {login_info.verification_url}")
+        while True:
```

Comment:
I think this needs to be a limited loop. If it fails to get login after some finite amount of time, it should just fail.

Reply:
solved in [de353c7](https://github.com/omnivector-solutions/jobbergate/pull/228/commits/de353c7382c3a765f9954042b2776d5fa716744c).

Do you feel this is enough?